README.md
# freebsd-ci

This is a tool to implement a local CI/CD server for FreeBSD.  
You'll need a FreeBSD system (a VM is fine) and a Github account
## Installation

You can install `freebsd-ci` on a FreeBSD machine with a rust toolchain installed and a ZFS pool; to use `freebsd-ci`, you'll need root access.

**NOTE**: the FreeBSD machine has to be the same verison of more recent than the container you want to use.  
A FreeBSD 12.0 system can run a FreeBSD 11.2 container. A FreeBSD 11.2 system can not run a FreeBSD 12.0 container.

You can install rust via rustup (https://rustup.rs) or installing the package via `pkg`

To install `freebsd-ci` you can clone this repo in your directory and then run:
```console
# cargo install --path .
```

FreeBSD 12.0 and the `latest` pkg repository are highly suggested.

The `freebsd-ci` tool is based on `pot` containers. So to install it, you can run:

```console
# pkg install pot
# vi /usr/local/etc/pot/pot.conf # check the configuration
# pot init
```

### Create your images catalog

The most tricky part is to provide an images catalog, that will be used to build software projects.  
In the `pot-images` folder, there are some scripts, ready to use.  
To use them, you can copy them in the pot flavor folder:
```console
# cp pot-images/bsd-ci-rust-* /usr/local/etc/pot/flavours/
# pot ls -F
flavour: bsd-ci-rust-nightly
flavour: bsd-ci-rust-stable
flavour: dns
```

To create the rust stable image for FreeBSD 11.2, use the command:
```console
# pot create -p FreeBSD-11_2-rust-stable -b 11.2 -t single -f bsd-ci-rust-stable
```

To create the rust nightly image for FreeBSD 12.0, use the command:
```console
# pot create -p FreeBSD-12_0-rust-nightly -b 12.0 -t single -f bsd-ci-rust-nightly
```

You can create all the combinations you need.  
Be aware that the name of the pot (the `-p` option) has a fixed form, in order to be found by the freebsd-ci tool.  
You can list your `pot`, via the command:
```console
# pot ls
```

**NOTE**: `pot` names cannot contains dots; for this reason, the dot is subsituted by an underscore.
### The freebsd-ci.conf

The `freebsd-ci.conf` file is needed to store your github token. It's a simple toml file:
```toml
[tokens]
github = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
```

Github tokens can be obtained at the url https://github.com/settings/tokens/new and you need the `repo` and the `user` scope.


## How to use it

The command line tools is called `freebsd-ci` and how to use it can be invoked in this way:

```console
# freebsd-ci --help
USAGE:
    freebsd-ci [FLAGS] [OPTIONS] --project <project_name> --user-name <user_name>

FLAGS:
    -f, --force                A Flag to force operations (i.e. remove fscomp or images with the same name)
    -h, --help                 Prints help information
    -B, --build-script-only    A Flag to rendert the build script only (on stdout)
    -v, --verbose              Enable the verbose output No multiple occurrences are supported
    -V, --version              Prints version information

OPTIONS:
    -b, --build <build_template>    The pathname to the build-sh template [default: ./templates/build.sh]
    -c, --config <configfile>       The pathname to the toml configuration file [default: ./freebsd-ci.conf]
    -P, --project <project_name>    Github project name
    -T, --tag-name <tag_name>       Tag name: Using this option, a tag will be built. If a related release is found, the
                                    artifacts will be uploaded
    -U, --user-name <user_name>     Github user name
```
where `username` is the github username and `project-name` is the github project name and are manddatory.

To test that you installation works, from the project directory, you can try to build my test project:
```
# freebsd-ci -U pizzamig -P ci-test
```

### The build.sh template
The build script template can be customized. in `templates/build.sh` there is a standard script with all template variables listed and documented.  
If you want to test your script template you can use the `-b` option to point to your custom template and the flag -B that will show the output at the console, without executing the build (the project will be still downloaded to read the YAML file)

### The deploy to github

If the tool is invoked with the `-T` option, then the a tarball can be built and uploaded to github to the relative release.  
The upload is performed by the build script and it can be disabled in the YAML file

### The YAML file

The YAML file has to be stored in the root directory with the name `.bsd-ci.yml`.  
Here a commented example:
```yaml
os: FreeBSD		# the operating system

FreeBSD:		# the version of the operating system
    - '11.2'
    - '12.0'

update: true	# perform the toolchain/pkg upgrade before the build

language: rust	# the project language

rust:			# the language variant
    - stable
    - beta
    - nightly

no_deploy:		# in case of a tag (deploy) which version to exclude
    rust:
        - nightly
        - beta
```






CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.1] 2019-02-01
### Fixed
- -b option wasn't honored

## [0.1.0] 2019-02-01
### Added
- github support
- build.sh: using template support via tera
- build.sh: add template context
- add a --build-script-only flag, to render the build.sh script
- add a --tag-name option, to support automatic asset upload
- add support to custom build template scripts
- add support to the no_deploy clause in the yaml file
- asset support: delete the asset, if already present, allowing "update"


src/main.rs
mod builder;
mod config;
mod error;
mod github;
mod pot;
mod yaml;
use crate::builder::build;
use crate::error::ParseError;
use crate::github::{get_release_id, get_status, AssetJson};
use crate::yaml::{
    get_build_lang, get_build_os, get_lang, get_no_deploy, get_os, get_update, get_yaml,
};
use exitfailure::ExitFailure;
use failure::ResultExt;
use log::{debug, error, info};
use std::fmt::Display;
use std::path::PathBuf;
use std::string::ToString;
use structopt::StructOpt;
use yaml_rust::YamlLoader;

#[derive(Debug, StructOpt)]
struct Opt {
    #[structopt(flatten)]
    verbose: structopt_flags::SimpleVerbose,
    /// The pathname to the toml configuration file
    #[structopt(
        short = "-c",
        long = "--config",
        parse(from_os_str),
        default_value = "./freebsd-ci.conf"
    )]
    configfile: PathBuf,
    /// The pathname to the build-sh template
    #[structopt(
        short = "-b",
        long = "--build",
        parse(from_os_str),
        default_value = "./templates/build.sh"
    )]
    build_template: PathBuf,
    /// A Flag to force operations (i.e. remove fscomp or images with the same name)
    #[structopt(short = "-f", long = "--force")]
    force_flag: bool,
    /// A Flag to rendert the build script only (on stdout)
    #[structopt(short = "-B", long = "--build-script-only")]
    render_build_flag: bool,
    /// Github project name
    #[structopt(short = "-P", long = "--project")]
    project_name: String,
    /// Github user name
    #[structopt(short = "-U", long = "--user-name")]
    user_name: String,
    /// Tag name: Using this option, a tag will be built. If a related release is found,
    /// the artifacts will be uploaded
    #[structopt(short = "-T", long = "--tag-name")]
    tag_name: Option<String>,
}

#[derive(Debug, Clone)]
struct BuildLang {
    lang: String,
    lang_variant: String,
}

impl Display for BuildLang {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        writeln!(
            f,
            "BuildLang {} (variant: {})",
            self.lang, self.lang_variant
        )
    }
}

#[derive(Debug, Clone)]
struct BuildOS {
    os_family: String,
    os_version: String,
}

impl Display for BuildOS {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        writeln!(f, "BuildOS {} {}", self.os_family, self.os_version)
    }
}

#[derive(Debug)]
struct BuildJob {
    lang: BuildLang,
    os: BuildOS,
    deploy: bool,
}

fn print_jobs(jv: &[BuildJob]) {
    println!("\nJob queue of {} elements", jv.len());
    for j in jv {
        println!(
            "BuildJob: {} {}  {} ({}) deploy: {}",
            j.os.os_family, j.os.os_version, j.lang.lang, j.lang.lang_variant, j.deploy
        );
    }
}

impl ToString for BuildJob {
    fn to_string(&self) -> String {
        let mut rc = self.os.os_family.clone();
        rc.push('-');
        rc.push_str(&self.os.os_version);
        rc.push('-');
        rc.push_str(&self.lang.lang);
        rc.push('-');
        rc.push_str(&self.lang.lang_variant);
        rc.replace(".", "_")
    }
}

#[derive(Debug)]
pub(crate) struct Project {
    pub(crate) owner: String,
    pub(crate) project: String,
}

impl ToString for Project {
    fn to_string(&self) -> String {
        format!("{}__{}", self.owner, self.project)
    }
}

#[derive(Debug, Default)]
pub(crate) struct BuildOpt {
    pub(crate) update: bool,
    pub(crate) release_id: Option<u64>,
    pub(crate) assets: Vec<AssetJson>,
}

fn main() -> Result<(), ExitFailure> {
    let opt = Opt::from_args();
    env_logger::try_init()?;
    debug!("BSD Continuous integration");
    /* Initial checks */
    if !crate::pot::is_pot_available() {
        error!("This error needs better explanation");
        return Ok(());
    }
    debug!(
        "Reading configuration file {}",
        opt.configfile
            .to_str()
            .unwrap_or("Filename not convertible")
    );
    let config = crate::config::get_config(&opt.configfile).with_context(|_| {
        format!(
            "could not parse the file {}",
            opt.configfile.to_str().unwrap_or("file name not printable")
        )
    })?;
    let prj = Project {
        owner: opt.user_name.clone(),
        project: opt.project_name.clone(),
    };
    let (rs, _) = get_status(&prj, &config.tokens.github).with_context(|_| {
        format!(
            "Fecth repository data failed for user {} project {}",
            prj.owner, prj.project,
        )
    })?;
    println!("github repository information:\n{}", rs);
    /* fetch the repo to read the .bsd-ci file */
    let path = crate::pot::fetch_git_in_fscomp(&prj, &rs, &opt).with_context(|_| {
        "Failed to create a ZFS dataset with the project in it\n Is pot installed?\n Are you root?"
    })?;

    println!("Git repo fetched in {}", path);

    let mut build_queue = Vec::new();
    let mut build_opt = BuildOpt::default();
    let yaml_string =
        get_yaml(&path).with_context(|_| "Error accessing the yaml file in the project")?;
    let docs = YamlLoader::load_from_str(&yaml_string)
        .with_context(|_| "Error parsing the yaml file in the project")?;
    for d in docs {
        let h = d.into_hash().unwrap();
        debug!("{:?}", h);
        let lang = get_lang(&h).with_context(|_| "Invalid YAML")?;
        let os = get_os(&h).with_context(|_| "Invalid YAML")?;
        let build_lang = match lang.as_ref() {
            "rust" => get_build_lang("rust", &h)
                .with_context(|_| format!("Ivalid YAML for language {}", lang))?,
            _ => {
                return Err(ExitFailure::from(ParseError::GenericError {
                    msg: "language not supported".to_string(),
                }));
            }
        };
        let build_os = match os.as_ref() {
            "FreeBSD" => get_build_os("FreeBSD", &h)?,
            _ => {
                return Err(ExitFailure::from(ParseError::GenericError {
                    msg: "os not supported".to_string(),
                }));
            }
        };
        build_opt.update = get_update(&h)?;
        info!("{:?}", build_lang);
        info!("{:?}", build_os);
        for o in &build_os {
            for l in &build_lang {
                build_queue.push(BuildJob {
                    lang: l.clone(),
                    os: o.clone(),
                    deploy: true,
                });
                debug!("o {:?} - l {:?}", o, l);
            }
        }
        get_no_deploy(&h, &mut build_queue)?;
        print_jobs(&build_queue);
    }
    if let Some(tag_name) = &opt.tag_name {
        if let Ok((release_id, assets, _)) = get_release_id(&prj, tag_name, &config.tokens.github) {
            build_opt.release_id = Some(release_id);
            build_opt.assets = assets;
        }
    }
    build(&build_queue, &prj, &opt, &build_opt, &config.tokens.github)?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_job_to_string_rust() {
        let uut = BuildJob {
            os: BuildOS {
                os_family: "FreeBSD".to_string(),
                os_version: "11.2".to_string(),
            },
            lang: BuildLang {
                lang: "rust".to_string(),
                lang_variant: "stable".to_string(),
            },
            deploy: true,
        };
        let rc = uut.to_string();
        assert_eq!(&rc, "FreeBSD-11_2-rust-stable");
    }
    #[test]
    fn test_job_to_string_php() {
        let uut = BuildJob {
            os: BuildOS {
                os_family: "FreeBSD".to_string(),
                os_version: "12.0".to_string(),
            },
            lang: BuildLang {
                lang: "php".to_string(),
                lang_variant: "7.3".to_string(),
            },
            deploy: false,
        };
        let rc = uut.to_string();
        assert_eq!(&rc, "FreeBSD-12_0-php-7_3");
    }

    #[test]
    fn test_project_to_string_ok() {
        let uut = Project {
            owner: "pizzamig".to_string(),
            project: "potnet".to_string(),
        };
        let rc = uut.to_string();
        assert_eq!(&rc, "pizzamig__potnet");
    }
}
